pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,208 @@ package heal
 
 import (
 	"context"
+	"sync"
+	"time"
 
+	"github.com/edwarnicke/serialize"
 	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
-
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
 )
 
-// RegisterClientFunc - required to inform heal server about new client connection and assign it to the connection ID
-type RegisterClientFunc func(context.Context, *networkservice.Connection, networkservice.MonitorConnectionClient)
-
 type healClient struct {
-	ctx context.Context
-	cc  networkservice.MonitorConnectionClient
+	ctx              context.Context
+	cc               networkservice.MonitorConnectionClient
+	initOnce         sync.Once
+	initErr          error
+	connCache        map[string]*networkservice.Connection
+	conCacheExecutor serialize.Executor
+	expectedConns    verificationRequestMap
 }
 
 // NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
 func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) networkservice.NetworkServiceClient {
 	return &healClient{
-		ctx: ctx,
-		cc:  cc,
+		ctx:       ctx,
+		cc:        cc,
+		connCache: make(map[string]*networkservice.Connection),
 	}
 }
 
 func (u *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	u.initOnce.Do(func() {
+		errCh := make(chan error, 1)
+		pushFunc := healRequestFunc(ctx)
+		go u.listenToConnectionChanges(pushFunc, request.GetConnection().GetCurrentPathSegment(), errCh)
+		u.initErr = <-errCh
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
+
+	verificationRequest := u.addToExpectedConnections(request.GetConnection())
+	if verificationRequest != nil {
+		defer close(verificationRequest.stopCh)
+	}
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
```

Comment:
We probably want to use here updated path from `connCache`.

Reply:
I don't think so.
When Request() is called on client, it gets fresh path from previous chain elements.
If we replace it with the path we have, that path may be outdated.